### ISSUE TITLE
Implement replacement for DeferredList which doesn't subclass Deferred, and is cancellable

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4977]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#4977|
|Type|defect|
|Created|2011-03-22 16:27:16Z|

`DeferredList` doesn't override any functionality, and it has no reason to be a subclass (except for compatibility, jeez guys).

Composition > Inheritance.

The functionality `DeferredList` provides should be a free function that simply waits on a list of Deferreds and returns a regular Deferred.  For compatibility, there will still need to be a subclass with the whole mess of public attributes that it has, but we can eventually remove those.

This new code should also support cancellation from the beginning (hopefully a compelling reason, aside from the warnings, for developers to update).

<details><summary>Searchable metadata</summary>

```
trac-id__4977 4977
type__defect defect
reporter__glyph glyph
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__new new
resolution__None None
component__core core
keywords__ 
time__1300811236000000 1300811236000000
changetime__1300890921000000 1300890921000000
version__None None
owner__ 
cc__detly
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Why do you want to fix `DeferredList` in-place?  Let's add a lot of things like `gatherResults` that cover all of its functionality and then deprecate `DeferredList`.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_1):
> Why do you want to fix `DeferredList` in-place?  Let's add a lot of things like `gatherResults` that cover all of its functionality and then deprecate `DeferredList`.

I don't.  The "subclass with a whole mess of public attributes" that I was referring to would be called `DeferredList`.

Would it be compatible to just make `gatherResults` return a `Deferred` which isn't a `DeferredList`?  I think we might have to deprecate that as well and give the good version a new name.

While we're at it, this new code should return a cancellable Deferred which will cancel all of the things that it's still waiting on.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

> Would it be compatible to just make gatherResults return a Deferred which isn't a DeferredList?

I think so.  However, `gatherResults` has other problems (mostly in `_parseDListResult` - `__debug__`?  `assert`?  _noooooo..._), so we may want a new API anyway.
